### PR TITLE
Implement Go servlet

### PR DESCRIPTION
**Pull request by tttzach**
Initial implementation of the Go servlet.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        streetAddressWaypointsAsCoordinates,
+        nonStreetAddressWaypointsAsPlaceTypes);
+
+    // All street address coordinates including origin and destination are collected
+    List<LatLng> streetAddressesAsCoordinates = streetAddressWaypointsAsCoordinates;
+    streetAddressesAsCoordinates.add(originAsCoordinates);
+    streetAddressesAsCoordinates.add(destinationAsCoordinates);
+
+    // Remove all null entries of street address coordinates and non street address waypoints
+    streetAddressesAsCoordinates =
```
Comment (austinteshuba) [inline src/main/java/com/google/sps/servlets/GoServlet.java:279]: Why would there be null entries here? I would add a comment explaining what the represent, and possibly try to eliminate them earlier on (i.e. if they are being added from the `seperateWaypoints` method, I would advise adding a null check in that method to avoid this issue, and if this is a result of `originAsCoordinates` or `destinationAsCoordinates` being null, I would avoid adding them to begin with if they are null and possibly return early if this results in failure)

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+   */
+  public List<List<String>> searchNearbyEveryKnownLocationForClosestPlaceTypeMatch(
+      List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes,
+      List<LatLng> streetAddressesAsCoordinates)
+      throws PlacesException {
+    List<List<String>> allSearchNearbyResults = new ArrayList<>();
+    for (PlaceType nonStreetAddressWaypoint : nonStreetAddressWaypointsAsPlaceTypes) {
+      List<String> searchNearbyResults = new ArrayList<>();
+      for (LatLng coordinate : streetAddressesAsCoordinates) {
+        PlaceType placeType = nonStreetAddressWaypoint;
```
Comment (austinteshuba) [inline src/main/java/com/google/sps/servlets/GoServlet.java:196]: No need for the variable `placeType`

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+      String origin, String destination, List<List<String>> allWaypointCombinations)
+      throws DirectionsException {
+    long minTravelTime = 0;
+    List<String> mostOptimalWaypointCombination = new ArrayList<String>();
+    for (List<String> waypointCombination : allWaypointCombinations) {
+      DirectionsClient directionsClient = directionsClientFactory.getDirectionsClient(apiKey);
+      DirectionsResult directionsResult =
+          directionsClient.getDirections(origin, destination, waypointCombination);
+      long travelTime = DirectionsClient.getTotalTravelTime(directionsResult);
+      if (minTravelTime == 0 || travelTime < minTravelTime) {
```
Comment (austinteshuba) [inline src/main/java/com/google/sps/servlets/GoServlet.java:230]: Use `Long.MAX_VALUE` for the initial value of `minTravelTime`, and then you can simplify this check to be `Math.min(minTravelTime, travelTime)`

File: src/test/java/GoServletTest.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  private DirectionsRoute shorterRoute = new DirectionsRoute();
+  private DirectionsResult shorterResult = new DirectionsResult();
+
+  @Override
+  @Before
+  public void setUp() throws Exception {
+    super.setUp();
+    directionsClientFactory = Mockito.mock(DirectionsClientFactory.class);
+    directionsClient = Mockito.mock(DirectionsClient.class);
+    placesClientFactory = Mockito.mock(PlacesClientFactory.class);
```
Comment (austinteshuba) [inline src/test/java/GoServletTest.java:84]: You shouldn't have to mock a factory

File: src/test/java/GoServletTest.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  private DirectionsClient directionsClient;
+  private PlacesClientFactory placesClientFactory;
+  private PlacesClient placesClient;
+  private TasksClientFactory tasksClientFactory;
+  private TasksClient tasksClient;
+  private GeocodingClientFactory geocodingClientFactory;
+  private GeocodingClient geocodingClient;
+  private GoServlet servlet;
+
+  private static final String API_KEY = "fake api key";
```
Comment (austinteshuba) [inline src/test/java/GoServletTest.java:57]: nit: Keep `private` and `private static` variables together (i.e. move the `private` declarations up)

File: src/test/java/GoServletTest.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    tasksClient = Mockito.mock(TasksClient.class);
+    geocodingClientFactory = Mockito.mock(GeocodingClientFactory.class);
+    geocodingClient = Mockito.mock(GeocodingClient.class);
+
+    streetAddressGeocodingResult.geometry = streetAddressGeometry;
+    streetAddressGeocodingResult.geometry.location = streetAddressCoordinates;
+    streetAddressGeocodingResult.types = new AddressType[] {AddressType.STREET_ADDRESS};
+    restaurantGeocodingResult.types = new AddressType[] {AddressType.RESTAURANT};
+
+    longerDuration.inSeconds = 10;
```
Comment (austinteshuba) [inline src/test/java/GoServletTest.java:96]: Maybe set the time value to be a constant

File: src/test/java/GoServletTest.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  private static final String RESTAURANT_TWO = "restaurant near coordinate two";
+  private static final String BANK_ONE = "bank near coordinate one";
+  private static final String BANK_TWO = "bank near coordinate two";
+
+  private GeocodingResult streetAddressGeocodingResult = new GeocodingResult();
+  private Geometry streetAddressGeometry = new Geometry();
+  private LatLng streetAddressCoordinates = new LatLng(0, 0);
+  private GeocodingResult restaurantGeocodingResult = new GeocodingResult();
+
+  private Duration longerDuration = new Duration();
```
Comment (austinteshuba) [inline src/test/java/GoServletTest.java:68]: Any reason why these are not static? Seems like they are initialized with the same values each time

File: src/test/java/GoServletTest.java [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+            directionsClientFactory,
+            placesClientFactory,
+            tasksClientFactory,
+            geocodingClientFactory,
+            API_KEY);
+
+    Mockito.when(directionsClientFactory.getDirectionsClient(API_KEY)).thenReturn(directionsClient);
+    Mockito.when(placesClientFactory.getPlacesClient(API_KEY)).thenReturn(placesClient);
+    Mockito.when(tasksClientFactory.getTasksClient(Mockito.any())).thenReturn(tasksClient);
+    Mockito.when(authenticationVerifier.verifyUserToken(Mockito.anyString()))
```
Comment (austinteshuba) [inline src/test/java/GoServletTest.java:117]: Shouldn't have to specify the mock for `authenticationVerfier` - this should be handled by the super class

File: src/test/java/GoServletTest.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        .thenReturn(AUTHENTICATION_VERIFIED);
+    Mockito.when(geocodingClientFactory.getGeocodingClient(API_KEY)).thenReturn(geocodingClient);
+  }
+
+  @Test
+  public void separateWaypoints() throws Exception {
+    // Street address should be converted into coordinates, restaurant should be converted into
+    // place type
+    List<String> waypoints = ImmutableList.of("street address", "restaurant");
```
Comment (austinteshuba) [inline src/test/java/GoServletTest.java:126]: Make this a constant, since `streetAddressGeocodingResult` is already a constant. This list, as well as its contents, have semantic value

File: src/test/java/GoServletTest.java [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+            ImmutableList.of("place_id:" + BANK_ONE, "place_id:" + BANK_TWO)),
+        actual);
+  }
+
+  @Test
+  public void chooseWaypointCombinationWithShortestTravelTime() throws Exception {
+    // Find the waypoints which result in the shorter travel time
+    String origin = "A";
+    String destination = "B";
+    List<String> longerTravelTimeWaypoints = ImmutableList.of("long", "long");
```
Comment (austinteshuba) [inline src/test/java/GoServletTest.java:184]: Ditto on making this a constant

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+   * @throws ServletException
+   * @throws IOException
+   */
+  @Override
+  public void doGet(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws ServletException, IOException {
+    assert googleCredential != null
+        : "Null credentials (i.e. unauthenticated requests) should already be handled";
+    // Get all tasks from user's tasks account TODO: Get relevant tasks using task
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Added PR number since it's done already, thanks for pointing it out!

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+    // titles
+    TasksClient tasksClient = tasksClientFactory.getTasksClient(googleCredential);
+    DirectionsClient directionsClient = directionsClientFactory.getDirectionsClient(apiKey);
+
+    // Initialize Tasks Response
+    List<Task> tasks;
+
+    String taskLists = request.getParameter("taskLists");
+    if (taskLists == null) {
+      tasks = TasksServlet.getAllTasksFromAllTaskLists(tasksClient);
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Initially yes, but not anymore because these methods have been moved to a utility class in https://github.com/googleinterns/blueprint-step-2020/pull/146.

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+    this.placesClientFactory = placesClientFactory;
+    this.tasksClientFactory = tasksClientFactory;
+    this.geocodingClientFactory = geocodingClientFactory;
+    this.apiKey = apiKey;
+  }
+
+  /**
+   * Returns the most optimal order of travel between addresses.
+   *
+   * @param request HTTP request from the client.
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:103]: Oh good point, did not register query strings as parameters in my head. Thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+    } else {
+      List<String> selectedTaskListIds = Arrays.asList(taskLists.split(","));
+      tasks = TasksServlet.getAllTasksFromSpecificTaskLists(tasksClient, selectedTaskListIds);
+    }
+
+    String origin = request.getParameter("origin");
+    String destination = request.getParameter("destination");
+    List<String> waypoints = LocationsUtility.getLocations("Location", tasks);
+
+    try {
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:133]: Thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+   *     non street addresses.
+   * @param streetAddressWaypoints A pointer to the result for street address waypoints to achieve
+   *     the effect of returning multiple types at once.
+   * @param streetAddressWaypointsAsCoordinates A pointer to the result for street address waypoints
+   *     as coordinates to achieve the effect of returning multiple types at once.
+   * @param nonStreetAddressWaypointsAsPlaceTypes A pointer to the result for non street address
+   *     waypoints as place types to achieve the effect of returning multiple types at once.
+   * @throws GeocodingException An exception thrown when an error occurs with the Geocoding API.
+   */
+  public void separateWaypoints(
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:160]: Yeah I agree a lot is going on in this method and I would also like to know if @gbuenoandrade or @antmar has any ideas on how to break this (and the other large methods up). Unfortunately, it's not the looping that's expensive, it's the call to the Geocoding API at line 169.

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+    List<LatLng> streetAddressWaypointsAsCoordinates = new ArrayList<>();
+    List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes = new ArrayList<>();
+    separateWaypoints(
+        waypoints,
+        streetAddressWaypoints,
+        streetAddressWaypointsAsCoordinates,
+        nonStreetAddressWaypointsAsPlaceTypes);
+
+    // All street address coordinates including origin and destination are collected
+    List<LatLng> streetAddressesAsCoordinates = streetAddressWaypointsAsCoordinates;
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Hmm it would be misleading to add the origin and destination into a list that is meant for waypoints. I've sort of took that direct assignment out and used `addAll` instead. I don't know if that helps, let me know!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 1 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: The null entries are from the other PRs where you suggested empty optionals for. I've added an optional empty check for `originAsCoordinates` and `destinationAsCoordinates` to throw an error if that happens. Thanks for pointing it out!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 2 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:201]: Agreed, thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 3 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:235]: Hmm not sure if I am understanding correctly, I kinda need to reassign `mostOptimalWaypointCombination` so I don't think that would help.

File: src/test/java/GoServletTest.java (hunk 4 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:85]: Oh true, thanks for pointing that out!

Comment (tttzach) [reply src/test/java/GoServletTest.java:85]: Not sure if I am understanding correctly again, kinda need it to give explicit implementations to the GoServlet constructor. The default constructor gets API keys using file I/O.

File: src/test/java/GoServletTest.java (hunk 5 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:58]: Good idea, thanks!

File: src/test/java/GoServletTest.java (hunk 8 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java]: Oh yeah you're right, thanks!

File: src/test/java/GoServletTest.java (hunk 6 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:97]: Good call, thanks!

File: src/test/java/GoServletTest.java (hunk 7 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:69]: Uh nope, just forgot - thanks!

File: src/test/java/GoServletTest.java (hunk 10 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:186]: Good call, thanks!

File: src/test/java/GoServletTest.java (hunk 9 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:125]: Yep it's used twice too, thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:104]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:134]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 1 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 2 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/test/java/GoServletTest.java (hunk 5 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:73]: `RESOLVED`

File: src/test/java/GoServletTest.java (hunk 6 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java]: `RESOLVED`

File: src/test/java/GoServletTest.java (hunk 7 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java]: `RESOLVED`

File: src/test/java/GoServletTest.java (hunk 8 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java]: `RESOLVED`

File: src/test/java/GoServletTest.java (hunk 9 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java]: `RESOLVED`

File: src/test/java/GoServletTest.java (hunk 10 above)
Comment (tttzach) [reply src/test/java/GoServletTest.java:193]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 3 above)
Comment (gbuenoandrade) [reply src/main/java/com/google/sps/servlets/GoServlet.java:236]: It would remove the need to check if `mintravelTime == 0`. In competitive programming, we always do that. As a matter of fact, there's a magic number we use to represent infinity "0x3f3f3f3f". It is "magic" because it is large, doubling it still fits into a 32 bit integer, and because the bytes of it are the same, you can fill an array with that value using the C function `memset`. Though, I'm not sure if that's considered a "good" practice :( Maybe using `Optional` would be preferred.

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+    streetAddressesAsCoordinates.add(originAsCoordinates);
+    streetAddressesAsCoordinates.add(destinationAsCoordinates);
+
+    // Remove all optional empty entries of street address coordinates and non street address
+    // waypoints
+    List<LatLng> nonEmptyStreetAddressesAsCoordinates =
+        streetAddressesAsCoordinates.stream()
+            .filter(coordinates -> !coordinates.equals(Optional.empty()))
+            .map(coordinates -> coordinates.get())
+            .collect(Collectors.toList());
```
Comment (gbuenoandrade) [inline src/main/java/com/google/sps/servlets/GoServlet.java:295]: maybe create an auxiliary method to do that? It could accept `List<Optional<T>>` so you can use it for both lists

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+      Set<String> selectedTaskListIds = new HashSet<>(Arrays.asList(taskLists.split(",")));
+      tasks = TasksUtility.getAllTasksFromSpecificTaskLists(tasksClient, selectedTaskListIds);
+    }
+
+    String origin = request.getParameter("origin");
+    String destination = request.getParameter("destination");
+    List<String> waypoints = LocationsUtility.getLocations("Location", tasks);
+
+    try {
+      List<String> mostOptimalWaypointCombination =
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:135]: Agreed, thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (austinteshuba) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Yes, I think this makes it more readable. Thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 3 above)
Comment (austinteshuba) [reply src/main/java/com/google/sps/servlets/GoServlet.java:236]: @gbuenoandrade Sounds like a cool number :) +1 on using optional if `Long.MAX_VALUE` is not recommended

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:161]: Oh this is a great idea, thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+            GeocodingResultUtility.convertToPlaceType(waypoint));
+      }
+    }
+  }
+
+  /**
+   * Search nearby every street address with known coordinates for a place type match. For example,
+   * if we know exactly where our houses are and we are looking for a restaurant, we search for a
+   * restaurant closest to your house and a restaurant closest to my house. In this case, the method
+   * should return [[restaurantOne, restaurantTwo]]. Scope of method is public for testing purposes.
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:185]: Yeah I agree with you, I could have definitely moved that into a class on its own and abstracted out the API call. In the interest of time, I will just imprint that into my mind for the next internship. :(

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 20]
```diff
@@ -0,0 +1,316 @@
...
+  public List<List<String>> searchNearbyEveryKnownLocationForClosestPlaceTypeMatch(
+      List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes,
+      List<LatLng> streetAddressesAsCoordinates)
+      throws PlacesException {
+    List<List<String>> allSearchNearbyResults = new ArrayList<>();
+    for (PlaceType nonStreetAddressWaypoint : nonStreetAddressWaypointsAsPlaceTypes) {
+      List<String> searchNearbyResults = new ArrayList<>();
+      for (LatLng coordinate : streetAddressesAsCoordinates) {
+
+        PlacesClient placesClient = placesClientFactory.getPlacesClient(apiKey);
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Sadly, it doesn't work because it needs a new client each time otherwise it throws `java.lang.IllegalStateException: 'await', 'awaitIgnoreError' or 'setCallback' was already called.`. Same with the Geocoding client. :/

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 21]
```diff
@@ -0,0 +1,316 @@
...
+   *     restaurant, supermarket, police station)
+   * @param streetAddressesAsCoordinates A list of coordinates to look for the place types around.
+   * @return A list of lists of place IDs where each list represents the search nearby result for
+   *     every known coordinate.
+   * @throws PlacesException An exception thrown when an error occurs with the Places API.
+   */
+  public List<List<String>> searchNearbyEveryKnownLocationForClosestPlaceTypeMatch(
+      List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes,
+      List<LatLng> streetAddressesAsCoordinates)
+      throws PlacesException {
```
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:220]: True, thanks! That method name is a lot shorter.

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Awesome, thanks!

[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 3 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:258]: Optional is a great idea, I think it gives more information. Thanks!

Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:258]: minTravelTime.isPresent()

Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:258]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:208]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:220]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 17 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Gonna ask you on chat - not entirely sure how to set types in Java templates

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (gbuenoandrade) [reply src/main/java/com/google/sps/servlets/GoServlet.java:227]: Gotcha
RESOLVED

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 22]
```diff
@@ -0,0 +1,332 @@
...
+    // All street address coordinates including origin and destination are collected
+    List<Optional<LatLng>> streetAddressesAsCoordinates = new ArrayList<>();
+    streetAddressesAsCoordinates.addAll(separatedWaypoints.streetAddressWaypointsAsCoordinates);
+    streetAddressesAsCoordinates.add(originAsCoordinates);
+    streetAddressesAsCoordinates.add(destinationAsCoordinates);
+
+    // Remove all optional empty entries of street address coordinates and non street address
+    // waypoints
+    List<LatLng> nonEmptyStreetAddressesAsCoordinates = filterNonNull(streetAddressesAsCoordinates);
+    List<PlaceType> nonEmptynonStreetAddressWaypointsAsPlaceTypes =
```
Comment (gbuenoandrade) [inline src/main/java/com/google/sps/servlets/GoServlet.java:315]: nit: nonEmptyNonStreet

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 23]
```diff
@@ -0,0 +1,332 @@
...
+        searchForPlacesNearLocations(
+            nonEmptynonStreetAddressWaypointsAsPlaceTypes, nonEmptyStreetAddressesAsCoordinates);
+
+    List<List<String>> allWaypointCombinations =
+        LocationsUtility.generateCombinations(allSearchNearbyResults);
+
+    List<String> optimalWaypointCombination =
+        chooseWaypointCombinationWithShortestTravelTime(
+            origin, destination, allWaypointCombinations);
+    optimalWaypointCombination.addAll(separatedWaypoints.streetAddressWaypoints);
```
Comment (gbuenoandrade) [inline src/main/java/com/google/sps/servlets/GoServlet.java:328]: why is this necessary?

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (gbuenoandrade) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: nit: I think you could also rename the parameters and local variables, but it's up to you

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 22 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java:315]: Uh oh, thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 23 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Oh not sure why I left that outside the function. Thanks for catching that!

Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 22 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 17 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Fixed, thanks! Like you mentioned, was just missing a <T>

Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: I renamed them, I agree they don't have to be as long in this method. Thanks!

[the remark this answers is not in the document]
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/servlets/GoServlet.java [hunk 24]
```diff
@@ -209,21 +210,20 @@ public GoResponse separateWaypoints(List<String> waypoints) throws GeocodingExce
...
    * @throws PlacesException An exception thrown when an error occurs with the Places API.
    */
   public List<List<String>> searchForPlacesNearLocations(
-      List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes,
-      List<LatLng> streetAddressesAsCoordinates)
+      List<PlaceType> placeTypeWaypoints, List<LatLng> streetAddressWaypoints)
       throws PlacesException {
     List<List<String>> allSearchNearbyResults = new ArrayList<>();
-    for (PlaceType nonStreetAddressWaypoint : nonStreetAddressWaypointsAsPlaceTypes) {
+    for (PlaceType nonStreetAddressWaypoint : placeTypeWaypoints) {
```
Comment (gbuenoandrade) [inline src/main/java/com/google/sps/servlets/GoServlet.java:224]: Super nit: the for variable now looks a bit confusing

File: src/main/java/com/google/sps/utility/GeocodingResultUtility.java [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.Optional;
+
+/** Utility class to extract data from GeocodingResult objects. */
+public class GeocodingResultUtility {
+  /**
+   * Parses for the first coordinate which is of a street address type in a result from the
+   * Geocoding API. If no street address type results are found, the coordinates of the first result
+   * is returned.
+   *
+   * @param result A GeocodingResult returned from the Geocoding API.
```
Comment (austinteshuba) [inline src/main/java/com/google/sps/utility/GeocodingResultUtility.java:32]: Javadoc does not match the header here - change parameter name

Comment (tttzach) [reply src/main/java/com/google/sps/utility/GeocodingResultUtility.java]: Done, thanks!

File: src/main/java/com/google/sps/servlets/GoServlet.java (hunk 24 above)
Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: Agreed, thanks!

Comment (tttzach) [reply src/main/java/com/google/sps/servlets/GoServlet.java]: `RESOLVED`

File: src/main/java/com/google/sps/utility/GeocodingResultUtility.java (hunk 25 above)
Comment (tttzach) [reply src/main/java/com/google/sps/utility/GeocodingResultUtility.java]: `RESOLVED`
